text/0000-network-spec.md

```diff
@@ -20,40 +20,96 @@ This document should serve as reference for all the clients to implement network
+
+## Peer Manager
+
+### API
+
+- **Broadcast Message** (This will use broadcast idea from )
+- **Send Message to PeerID** (Message will be routed to known PeerID) This will use Find route to peer id from routing table. Messages can travel with a TTL, so they get dropped after lurking around the network. This also avoid loops when the view of the graph is not the same to all nodes.
+- **Send Direct Message PeerID** In this case PeerID must be an active peer.
+- **Ban PeerID**
+
+### Storing information
+
+- Active Peers
+- Banned peers
+- Peer information (This is a long term database where we store information of all peers we have ever met).
+
+## Sync Service
+
+This service will handle keeping the network view for every peer up to date using Broadcast mechanism. This will manage MST. When new edge is created this information is propagated through the network.
+
+Also there is a hard sync mechanism for two nodes, that is used when new connection between two nodes is established.  Periodically nodes will run this hard sync to become up to date. The period to hard sync is always large enough so that changes are allowed to get propagated through the spanning tree.
+If view of the network differ by a large margin nodes will send their full view of the graph. Useful when new nodes join for the first time.
+Otherwise they will use bloom filters to find differences.
+
+**AddEdge/RemoveEdge** messages should contains a timestamp/nonce to avoid situations when removal gets to a node before addition (though in practice this should be extremely rare). When adding a new edge, both nodes after having signature from the other peer start broadcasting this new edge. Use multiple source broadcast mechanism.
+
+## Connecting Service
+
+This service will handle active connections.
+
+- It will determine if an incoming connection should be established. When some node is trying to establish a new connection if peers slots is full we will ping peer with latest last message. If ping fails, we drop our connection and establish connection with new peer.
+- It will try to establish new connections in several situations:
+    - Bootstrapping (When joining the network for the first time). Nodes will try to learn new nodes running
+    - Network balancing. We will have high tolerance to support connections to nodes that need to be close to use (same shard or block producers). Though we will have this number capped in case
+    - Low number of outbound connections.
+- Remove some of the active connections if the number of active peer goes over some threshold. We allow number of active peers to be unbound in case we need to establish new connections because lack of outbound connections, or because we becoming part of new shard.
```

Comment:
I think totally unbounded number of connections is dangerous as can knock node into "too many files open" mode. Which is hard to recover from as our storage layer also uses files and starts to fall apart.

Reply:
This is how I imagine it. Each connection have some of this properties (maybe more than once at the same time).
- Node is block/chunk producer with us.
- Is outbound connection.
- Is important regarding topology efficiency (edge in the MST).
- Is trusted (this is ambiguous it can be hardcoded/old/reliable connection).

We can fix a maximum number per property (notice again that several connection can share some properties). When a connection drops some properties and become not meaningful to us we can stop it gracefully. I think it would be very good if we don't cap connections with first property (though it might be an issue as you mentioned).

Reply:
This is how I imagine it. Each connection have some of this properties (maybe more than once at the same time).
- Node is block/chunk producer with us.
- Is outbound connection.
- Is important regarding topology efficiency (edge in the MST).
- Is trusted (this is ambiguous it can be hardcoded/old/reliable connection).

We can fix a maximum number per property (notice again that several connection can share some properties). When a connection drops some properties and become not meaningful to us we can stop it gracefully. I think it would be very good if we don't cap connections with first property (though it might be an issue as you mentioned).